docs/best-practices.md

```diff
@@ -0,0 +1,799 @@
+
+```rust
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Contract {
+    pub metadata: LazyOption<Metadata>,
+}
+
+#[derive(Serialize, Deserialize, BorshDeserialize, BorshSerialize)]
+#[serde(crate = "near_sdk::serde")]
+pub struct Metadata {
+    data: String,
+    image: Base64Vec,
+    blobs: Vec<String>,
+}
+
+#[near_bindgen]
+impl Contract {
+    #[init]
+    pub fn new(metadata: Metadata) -> Self {
+        Self {
+            metadata: LazyOption::new(b"m", Some(metadata)),
+        }
+    }
+
+    pub fn get_metadata(&self) -> Metadata {
+        // `.get()` reads and deserializes the value from the storage. 
+        self.metadata.get().unwrap()
+    }
+}
+```
+
+## Compile smaller binaries
+
+When compiling a contract make sure to pass flag `-C link-arg=-s` to the rust compiler:
+
+```bash
+RUSTFLAGS='-C link-arg=-s' cargo build --target wasm32-unknown-unknown --release
+```
+
+Here is the parameters we use for the most examples in `Cargo.toml`:
+
+```toml
+[profile.release]
+codegen-units = 1
+opt-level = "s"
+lto = true
+debug = false
+panic = "abort"
+overflow-checks = true
+```
+
+You may want to experiment with using `opt-level = "z"` instead of `opt-level = "s"` to see if generates a smaller binary. See more details on this in [The Cargo Book Profiles section](https://doc.rust-lang.org/cargo/reference/profiles.html#opt-level).
+
+## Use simulation testing
+
+**Note**: simulation testing is deprecated in favor of [Sandbox testing](https://github.com/near/sandbox).
+
+Simulation testing allows you to run tests for multiple contracts and cross-contract calls in a simulated runtime environment.
+Read more, [near-sdk-sim](https://github.com/near/near-sdk-rs/tree/master/near-sdk-sim)
+
+## Appendix
+
+### The traditional way of handling unique prefixes for persistent collections
+
+Hardcoded prefixes in the constructor using a short one letter prefix that was converted to a vector of bytes.
+When using nested collection, the prefix must be constructed manually.
+
+```rust
+#[near_bindgen]
+impl Contract {
+    #[init]
+    pub fn new() -> Self {
+        Self {
+            accounts: UnorderedMap::new(b"a"),
+            tokens: LookupMap::new(b"t"),
+            metadata: LazyOption::new(b"m"),
+        }
+    }
+
+    fn get_tokens(&self, account_id: &AccountId) -> UnorderedSet<String> {
+        let tokens = self.accounts.get(account_id).unwrap_or_else(|| {
+            // Constructing a unique prefix for a nested UnorderedSet.
+            let mut prefix = Vec::with_capacity(33);
+            // Adding unique prefix.
+            prefix.push(b's');
+            // Adding the hash of the account_id (key of the outer map) to the prefix.
+            // This is needed to differentiate across accounts.
+            prefix.extend(env::sha256(account_id.as_bytes()));
+            UnorderedSet::new(prefix)
+        });
+        tokens
+    }
+}
+```
```

Comment:
Our nesting section is empty, should this not just be put there https://sdk-docs-pr-37.onrender.com/contract-structure/nesting ?

Reply:
good idea, I moved this and the BorshStorageKey section to the nested collection page https://github.com/near/sdk-docs/pull/37/commits/360a5a1b1462ba18c90bb10b4b0589d0c6825c8e